Form layout—Documentation (HDS-4886) (#2947)
hashicorp/design-system: website/docs/components/form/layout/partials/guidelines/guidelines.md

> ## Form Header

Author: @LilithJames-HDS Added a section for the Form component above the Form Header component. (You could add the suggested edit above first and then tweak the text or add more content to it if you think it needs it.)

The Form component serves as the container for all Form layout components and form content. It ensures a consistent and streamlined user experience. 

### When to use

- To structure and organize form elements and inputs into a consistent layout
- When creating a form of any size in our products, including forms nested within Flyout or Modal components

### When not to use

- To organize non-form content
- To create a form for key-value pairs, use the [Key Value Inputs](/components/forms/key-value-inputs) component

> ![The anatomy of the Form layout, indicating the header, sections, separator, and actions](/assets/components/form/layout/form-anatomy.png)

Reviewer: Ah sorry I misunderstood! I fixed this now.

## Single column layout

The Form component ensures a consistent and streamlined user experience through the use of a single-column layout. Multi-column layouts can disrupt the visual flow and reading orderof fields which can result in missed inputs and increased friction. Multiple fields can be places in a row only when grouping related inputs improves the user experience. See [Section Multi Field Group](/components/form/layout#form-section-multi-field-group)

> ![The anatomy of the Form layout, indicating the header, sections, separator, and actions](/assets/components/form/layout/form-anatomy.png)

Reviewer 2: @LilithJames-HDS this image is dupliucate (not sure which one you want to keep)

## Spacing

Spacing in forms is based on a decreasing scale on an 8px grid. From largest to smallest, the spacing structure is:

> - Sections are 32px apart

Author: @LilithJames-HDS

> - Sections are 32px apart

Author: I submitted related suggestions for these: https://github.com/hashicorp/design-system/pull/2947/files#r2164646621

Resolving this thread to clean up

![Section Header anatomy showing a Title, a Description, and a placeholder for additional content](/assets/components/form/layout/form-section-anatomy.png)

### Section Field Group

The Section Field Group is used to lay out closely related form inputs in a single row and also sets tighter spacing between them to reinforce their relationship. (16px spacing vs. 24px)

In narrower screens, such as mobile devices, form inputs contained in a Section Field Group automatically “stack.”

![Section Field Group example depicting spacing between grouped fields](/assets/components/form/layout/form-section-fieldgroup.png)

## Separator

A Separator can be placed between Sections when a clearer visual distinction between Sections is needed.

!!! Do

A 32px gap is included above and below to divide the Separators and to avoid the appearance of being from the surrounding form elements; otherwise, it can appear “attached” to a specific section or field.

![Separator used to add extra visual differentiation between two Sections](/assets/components/form/layout/divider-spacing-do-1.png)

> !!!

Reviewer: Do you think having the template component is enough to not need to tell people about the 32px spacing? I think one of the big inconsistencies we were seeing was having Sections separated by Separators with variable gap settings (like using 24px instead of 32px)

> !!!

Reviewer 3: It's being communicated visually with the example and also in the Spacing section, so it feels like you have it covered. 

But if you're concerned about that specifically, I'd consider changing this into a "dont" with an example that has inconsistent section spacing.

!!! Dont

Don’t use a Separator between the last Section and the Button Set or form actions at the bottom of the Form content.

![Separator being uses after all the content in a form](/assets/components/form/layout/divider-spacing-dont.png)

!!!

## Form Footer

The final Section within a form is the Form Footer. The Form Footer typically houses a Button or [Button Set](/patterns/form-patterns#button-sets) to allow users to save or submit form information. For consistent spacing and layout, we recommend using a Button Set to contain the buttons/actions to Save/Submit or Cancel the form.

## Single column layout

The Form component ensures a consistent and streamlined user experience. We recommend using a single-column layout for forms, as multi-column layouts can disrupt the visual flow and reading order. Multi-column layouts can potentially result in missed inputs requiring resolution and thereby increasing friction. They should only be used when grouping related inputs improves the user experience.

> ![The anatomy of the Form layout, indicating the header, sections, separator, and actions](/assets/components/form/layout/form-anatomy.png)

Reviewer: Ah yes, I will update this

### Spacing

Spacing in forms is based on a decreasing scale on an 8px grid. From largest to smallest, the spacing structure is:

- Sections are 32px apart
- Separators are 32px from Section elements on either vertical side
- Inputs within a Section are 24px apart; unless they are within a Section Field Group
- Inputs within a Section Field Group have 16px spacing in wider viewports. In small viewports, the stacking matches the 24px spacing that all other Section elements share

> - Text elements inside the Form Header and Form Section Header are 8px apart

Reviewer 3: To be honest, this suggestion is quite confusing to me. I prefer they be more explicit.

> - Text elements inside the Form Header and Form Section Header are 8px apart

Author: I'm making suggestions of my own to simplify and clarify these.

The Form component serves as the container for all Form layout components and form content. It ensures a consistent and streamlined user experience. 

### When to use

- To structure and organize form elements and inputs into a consistent layout
- When creating a form of any size in our products, including forms nested within Flyout or Modal components

### When not to use

- To organize non-form content
- To create a form for key-value pairs, use the [Key Value Inputs](/components/forms/key-value-inputs) component

## Single column layout

> Multi-column layouts can disrupt the visual flow and reading order of fields which can result in missed inputs and increased friction. Multiple fields can be places in a row only when grouping related inputs improves the user experience. See [Section Multi Field Group](/components/form/layout#form-section-multi-field-group)

Author: ```suggestion
Multi-column form layouts can disrupt the visual flow and reading order of fields leading to increased friction and missed inputs. Multiple fields should be placed in a row only when grouping related inputs improves the user experience. See [Section Multi Field Group](/components/form/layout#form-section-multi-field-group)
```

![The anatomy of the Form layout, indicating the header, sections, separator, and actions](/assets/components/form/layout/form-anatomy.png)

## Spacing

Spacing in forms is based on a decreasing scale on an 8px grid. From largest to smallest, the spacing structure is:

- Sections are 32px apart

> - Separators are 32px from Section elements on either vertical side

Author: Mentioned above so this line can be deleted

- When inputs are stacked, they are 24px apart. When inputs are within a Section Field Group and side by side, e.g., in wider viewports, they are 16px apart.

> - Text elements inside the Form Header and Form Section Header are 8px apart

Author: ```suggestion
- Text elements within the Form Header and Form Section Headers are 8px apart.
```

![An Error Alert component used within a Form appearing above all the fields serving as a form-level error message](/assets/components/form/layout/form-layout-error.png)

## Form Section

Inputs within a form are organized in Sections. Multiple Sections may be used to group content for a more scannable and organized experience. Inputs that belong under the same heading or are related to the same topic should be within a Section.

### Form Section Header

Similar to the Form Header, the Section Header contains styled and spaced text elements, and a space for custom content to be used as needed.

![Section Header anatomy showing a Title, a Description, and a placeholder for additional content](/assets/components/form/layout/form-section-anatomy.png)

### Form Section Multi Field Group

The Section Multi Field Group organizes closely related form inputs in a single row, setting 16px between them, as opposed to 24px when stacked. This reinforces their relationship.

![Section Multi Field Group example depicting spacing between grouped fields](/assets/components/form/layout/form-section-fieldgroup.png)

> In narrower screens, such as mobile devices, form inputs contained in a Section Multi Field Group automatically stack at the ['md' breakpoint](/foundations/breakpoints). 

Author: There's no reason to use single quotes here.

!!! Do

Organize inputs together when they are closely related and beneficial to read together. No more than three inputs should be put within a single row.

![Three credit card information fields arranged with one full width field followed by a row of two side-by-side equal width fields](/assets/components/form/layout/form-do-rows.png)

!!!

!!! Dont

Place inputs in rows to save space or create columns within the form.

![Three equal width fields for credit card information arranged in a single row](/assets/components/form/layout/form-dont-rows.png)

!!!

## Form Separator

A Separator can be placed between Sections when a clearer visual distinction between Sections is needed.

!!! Do

Use a Separator to add space between Sections or before a new Section Header. 

![Separator used to add extra visual differentiation between two Sections](/assets/components/form/layout/divider-spacing-do-1.png)

!!!

!!! Dont

> Don’t use a Separator between the last Section and the Button Set or form actions at the bottom of the Form content.

Author: ```suggestion
Don’t use a Separator between the last Section and Form Footer / Button Set.
```

![Separator being uses after all the content in a form](/assets/components/form/layout/divider-spacing-dont.png)

!!!

## Form Footer

> The Form Footer typically houses a Button or Button Set to allow users to submit form information. When there's more than one action, we recommend using the [Button Set](/patterns/form-patterns#button-sets) component for consistent spacing and layout.

Author: ```suggestion
The Form Footer typically houses Buttons to allow users to submit form information. When there's more than one Button or action, we recommend using the [Button Set](/patterns/form-patterns#button-sets) component for consistent spacing and layout.
```

## Width and resizing behavior

### Max-width

> Sections have a default max-width of 672px, establishing a consistent line length throughout applications.

Author: ```suggestion
Sections have a default max-width of 672px, establishing a consistent line length throughout applications. It's possible to customize this max-width value if needed.
```

Author: We could also leave out this line: "It's possible to customize this max-width value if needed." since we already mention the max-width is a "default"

> ### When to use

Reviewer 3: ```suggestion
## Usage

### When to use
```

- Sections are 32px apart
- Separators are 32px from Section elements on either vertical side
- Inputs within a Section are 24px apart; unless they are within a Section Field Group
- Inputs within a Section Field Group have 16px spacing in wider viewports. In small viewports, the stacking matches the 24px spacing that all other Section elements share
- Text elements inside the Form Header and Form Section Header are 8px apart

![Representation of the various spacing sizes between Form layout components](/assets/components/form/layout/form-layout-spacing.png)

!!! Do

Organize inputs together when they are closely related and beneficial to read together. No more than three inputs should be put within a single row.

![Three credit card information fields arranged with one full width field followed by a row of two side-by-side equal width fields](/assets/components/form/layout/form-do-rows.png)

!!!

!!! Dont

Place inputs in rows to save space or create columns within the form.

![Three equal width fields for credit card information arranged in a single row](/assets/components/form/layout/form-dont-rows.png)

> !!!

Author: I don't we should repeat essentially the same long pieces of content in two places. I think it makes the content here for the Form Layout docs unnecessarily long, and it also makes the docs more difficult to maintain if there are updates in the future.

> !!!

Author: @LilithJames-HDS Sorry, I see this was resolved but from my understanding, @heatherlarsen said above that the "Single column layout" section would be ok to remove but to keep the "Spacing" section.

Personally, I would prefer to remove the "Single column layout" section to shorten the content and not repeat too much from the Pattern docs. (Also since the Form layout components don't provide a way to create true multi-column layouts so it doesn't seem very relevant.) I would just take this sentence and move it to the Multi Field group section: "Multiple fields should be placed in a row only when grouping related inputs improves the user experience."

(If you don't agree, feel free to leave as is. Just thought I would comment again in case there was a misunderstanding from before.)

> !!!

Reviewer 3: +1 to what @KristinLBradley said.

> !!!

Reviewer: It's duplicative of what we have in the Section Multi Field Group content so I think it's fine to remove it

> !!!

Author: Thanks for clarifying. @LilithJames-HDS

> Visit the [Form Patterns][/patterns/form-patterns] guidelines for additional details on spacing in forms.

Author: I can help add a new brief section for the Form component here above the Form Header section. (I recommend making the other recommended changes above first since they will help restructure other content coming before this.)

Author: Added a suggested edit to apply.

## Form Header

> An optional header is available within **the Form component**. The Header has two variants, one for the form-level header and one for section-level headers. Each variant contains styled and spaced text elements to serve as a starting point for consistent hierarchy and structure. All text elements have a gap of 8px applied.

Author: ```suggestion
An optional header is available within the **Form component**. The Header has two variants, one for the form-level header and one for section-level headers. Each variant contains styled text for a Title and Description to serve as a starting point. All Header child elements have a gap of 8px applied between them.
```

> The title and description can be customized, and a space for custom content can be edited or removed as needed.

Author: ```suggestion
The Title and Description content can be edited as needed. The space for custom content can also be edited or removed as needed.
```

![Form Header anatomy showing a Title, a Description, and a placeholder for additional content](/assets/components/form/layout/form-header-anatomy.png)

## Form-level error alert

Conventionally, input level validation is sufficient for error messaging within a form but overarching form-level messaging may be needed if validations happen on submission. Form-level error alerts should be applied in the case of form-level submission failures, timeouts during data submission, or when errors apply to multiple sections within the larger form.

The form-level error message will appear below the Header as its own Section and uses the Alert component.

![An Error Alert component used within a Form appearing above all the fields serving as a form-level error message](/assets/components/form/layout/form-layout-error.png)

## Form Section

Inputs within a form are organized in Sections. Multiple Sections may be used to group content for a more scannable and organized experience. Inputs that belong under the same heading or are related to the same topic should be within a Section.

### Section Header

Similar to the Form Header, the Section Header contains styled and spaced text elements.

> The title and description can be customized, and a space for custom content can be edited or removed.

Author: Btw, are you intending that the Title and Description text be required? They are optional in the Ember components (as I think they should be). So perhaps it would help to clarify in the guidelines that they can be deleted/removed if needed?

![Section Header anatomy showing a Title, a Description, and a placeholder for additional content](/assets/components/form/layout/form-section-anatomy.png)

> ### Section Field Groups

Author: ```suggestion
### Section Field Group
```

The Section Field Group is a way to organize closely related inputs, setting tighter spacing to visually reinforce their relation, e.g., address inputs.

> Section Field Group contents can be arranged horizontally or vertically and are spaced 16px apart, rather than the 24px between other inputs.

Author: This is no longer true and I included info on spacing in the other above suggested edit.

![Section Field Group example depicting spacing between grouped fields](/assets/components/form/layout/form-section-fieldgroup.png)

## Separator

A Separator can be placed between Sections when a clearer visual distinction between Sections is needed.

!!! Do

A 32px gap is included above and below to divide the Separators and to avoid the appearance of being from the surrounding form elements; otherwise, it can appear “attached” to a specific section or field.

> ![Separator used to add extra visual differentiation between two Sections](/assets/components/form/layout/divider-spacing-actions-do-1.png)

Author: This image is appearing as broken

https://hds-website-git-kristin-hds-4886-form-layout-d-efa4db-hashicorp.vercel.app/components/form/layout#separator

!!!

!!! Dont

> Don’t use Separators at the end of the last section between the fields and the button set or actions.

Author: ```suggestion
Don’t use a Separator between the last Section and the Button Set or form actions at the bottom of the Form content.
```

![Separator being uses after all the content in a form](/assets/components/form/layout/divider-spacing-dont.png)

!!!

## Button Set

> Buttons are necessary for communicating how the form details can be saved or submitted. We recommend using a [Button Set](/patterns/form-patterns#button-sets) so the user can Save/Submit or Cancel the form.

Author: ```suggestion
Buttons are necessary saving or submitting form information. For consistent spacing and layout, we recommend using a [Button Set](/patterns/form-patterns#button-sets) to contain the buttons/actions to Save/Submit or Cancel the form.
```

## Width and resizing behavior

### Max-width

Sections have a default max-width of 672px, establishing a consistent line length throughout applications.

> Where more control is needed, you can override the max-width property with any valid CSS value. In these cases, all sections should be set to the same max-width value.

Author: @LilithJames-HDS So for the Ember `Form` component, I ended up adding a `sectionMaxWidth` property that can be used to set a custom max-width value for the form content if desired.

> Where more control is needed, you can override the max-width property with any valid CSS value. In these cases, all sections should be set to the same max-width value.

Reviewer: That doesn't change the documentation, does it? Since we want that to be a backup. Or should the content change slightly to be 

```suggestion
Where more control is needed, you can override the max-width property using the `sectionMaxWidth` property. In these cases, all sections should be set to the same max-width value.
```

> Where more control is needed, you can override the max-width property with any valid CSS value. In these cases, all sections should be set to the same max-width value.

Reviewer: @heatherlarsen Would it be better to leave that note out of this tab completely in that case since it's not something we want people to use, it's a "just in case" feature?

> Where more control is needed, you can override the max-width property with any valid CSS value. In these cases, all sections should be set to the same max-width value.

Author: I think it would be fine to leave it out here since it's documented within the "Code" tab.
